### PR TITLE
Bugfix/cmake netcdf find

### LINKED ISSUE #656: CMake can't find NetCDF libraries on Cray XC
**Describe the bug**
On the Cray XC, CMake cannot find the NetCDF libraries.
I believe this is due to the nature of the library installations on the Cray.

The NetCDF installation directory contains a `bin` directory as expected, but not a `lib` or `include` directory.
Instead, the `lib` and `inlcude` directories are in separate subdirectories for each supported compiler (Cray, GNU, Intel, etc).
The cray compiler uses a wrapper script that sets the correct lib and include compiler flags based on the currently loaded programming environment.

However, this seems to be confusing CMake. If I set the `NetCDF_ROOT` env var relevant compiler sub-directory, CMake reports that it cannot run nc-config:

```
CMake Error at cmake/FindNetCDF.cmake:225 (message):
  Unable to properly find NetCDF.  Found static libraries at:
  /opt/cray/netcdf/4.6.1.3/CRAY/8.6/lib/libnetcdf.a but could not run
  nc-config:
Call Stack (most recent call first):
  model/src/CMakeLists.txt:102 (find_package)

CMake Error at cmake/FindNetCDF.cmake:225 (message):
  Unable to properly find NetCDF.  Found static libraries at:
  /opt/cray/netcdf/4.6.1.3/CRAY/8.6/lib/libnetcdff.a but could not run
  nc-config:
Call Stack (most recent call first):
  model/src/CMakeLists.txt:102 (find_package)
```

However, `nc-config` does exist, in the bin directory two levels up.

Furthermore, if CMake does manage to find nc-config, then running `nc-config --flibs` will not yield the correct result (it is empty in fact as there are 4 different sets of compile flags to chose from depending on the compiler selected).

My workaround for the moment is to disable the `find_package` call for NetCDF. Ideally, I don't want CMake to try and find the netCDF libraries and set the compiler flags - the compiler wrapper script handles all this so it is not technically required.

Perhaps I am misunderstanding something obvious (I am a bit of a cmake noob)?! So happy to be put right regarding my assumptions above!

Directory structure of my netCDF installation, in case it is useful:
```
netcdf/4.6.1.3
|-- bin
|-- CRAY
|   `-- 8.6
|       |-- include
|       `-- lib
|-- GNU
|   |-- 5.3
|   |   |-- include
|   |   `-- lib
|   `-- 7.1
|       |-- include
|       `-- lib
|-- INTEL
|   `-- 16.0
|       |-- include
|       `-- lib
```

**To Reproduce**
Run CMake on Cray XC with `NetCDF_ROOT=/opt/cray/netcdf/4.6.1.3/CRAY/8.6"
Relevant modules loaded:
 - cmake/3.21.3
 - PrgEnv-cray/5.2.82
 - cce/8.7.7(default)
 - cdt/18.12
 - cray-hdf5/1.10.2.0(default)
 - cray-netcdf/4.6.1.3(default)

**Expected behavior**
Should find netCDF library specific to currently loaded netCDF library and compiler environment, but fails with an error running nc-config.

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by ukmo-ccbunney**
# Pull Request Summary
Provides an update to CMake allowing it to avoid searching for a particular library (currently only netCDF).

## Description
Adds the `-D EXCLUDE_FIND="list"` option to CMakeLists.txt.

Specifying a library with `EXCLUDE_FIND` will stop CMake from searching for that library with `find_package`.

This is useful for systems where a compiler wrapper is used (such as the Cray) that automatically sets the compiler flags for library dependencies based on the current compiler environment loaded; in these cases CMake can fail to find the library or select an incorrect one.

The addition of the `CMAKE_OPTIONS` environment variable to `run_cmake_test`  is to allow extra options (like the `-DEXCLUDE_FIND`) to be passed to cmake from the matrix files.

No changes to regression tests expected.

### Issue(s) addressed
Fixes #656 

### Commit Message
Provides an update to CMake allowing it to avoid searching for a particular library 

### Check list  

- [x] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [x] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [n/a] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [n/a] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing
* How were these changes tested? Regression tests.
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?) Yes (no change expected)
* Have the matrix regression tests been run (if yes, please note HPC and compiler)? Cray XC HPC; GNU Fortran.

**Matrix results **
Usual known non-b4b tests (mww3_test_03 and ww3_tp2.10)
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (12 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (14 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
```

### CODE REVIEW

File: model/src/CMakeLists.txt [hunk 1]
```diff
@@ -90,12 +96,21 @@ if("OASIS" IN_LIST switches)
 endif()
 
 if(NETCDF)
-  find_package(NetCDF REQUIRED COMPONENTS Fortran)
+  if("netcdf" IN_LIST EXCLUDE_FIND)
+    # NetCDF library flags provided by compiler or wrapper script: don't search
+    set(NetCDF_Fortran_FOUND TRUE)
+	message(WARNING "Not searching for NetCDF library - please ensure correct flags are provided by your compiler/wrapper")
```
Comment (kgerheiser) [inline model/src/CMakeLists.txt]: Just a comment that I think this should be indented in line with `set`.

**Review by kgerheiser — APPROVED** [review]
Looks good

File: model/src/CMakeLists.txt (hunk 1 above)
Comment (ukmo-ccbunney) [reply model/src/CMakeLists.txt]: Ah yes - looks like a tab sneaked in there. Fixed now (in upcoming commit)

**Comment by ukmo-ccbunney** [comment]
I also need to fix the OASIS build for ww3_tp2.14 - this is suffering the same problem
I've made some changes, but regtest is currently segfaulting, so I need to investigate.

**Comment by kgerheiser** [comment]
OASIS is also supplied by the Cray compiler?

**Comment by ukmo-ccbunney** [comment]
> OASIS is also supplied by the Cray compiler?

No - but the CMake config for the oasis-mct build in ww3_tp2.14 suffers the same issues with NetCDF libraries.

**Comment by ukmo-ccbunney** [comment]
@mickaelaccensi  - Adding you as a reviewer as I changed a few lines in the OASIS-MCT build for ww3_tp2.14.

**Comment by aliabdolali** [comment]
@ukmo-ccbunney Thanks for adding DEXCLUDE_FIND to CMake options. Once @mickaelaccensi approves the PR, I will run it on our end, and will check the DEXCLUDE_FIND option on our RDHPC.

**Comment by mickaelaccensi** [comment]
I'm running it on my HPC, then I'll approve it

**Review by mickaelaccensi — APPROVED** [review]
it works on ifremer HPC

**Comment by aliabdolali** [comment]
> it works on ifremer HPC

Awesome, I just submitted the entire matrix on our hpc.

**Review by aliabdolali — APPROVED** [review]

**Comment by aliabdolali** [comment]
The test ran successfully on NOAA HPC with pre-known non-identical cases.

```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR2_UQ_MPI_d2                     (15 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (11 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (15 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (23 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (14 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (17 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (15 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_ufs1.3/./work_a                     (1 files differ)
```


[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/8542663/matrixDiff.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/8542664/matrixCompSummary.txt)
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/8542665/matrixCompFull.txt)

**Pull request merged**